Bugfixes, docstring improvements and add several unit tests for EELSModel class

magnunor opened the pull request:
This pull requests has several improvements for `EELSModel`:

- Add unit tests for several features which were previously not tested
- Fix bugs in 3 functions: `EELSModel.free_edges`, `EELSModel.enable_free_onset_energy` and `EELSModel.disable_free_onset_energy`
- Several docstring fixes, both typos, style fixes, and minor rewrites

ericpre reviewed (commented):
Can you be more specific in describing what does this PR fix? The git messages are generic and it would be good to put together the test and the fix in the same commit - it is useful when looking back at the git history to understand the reasoning of the changes.

magnunor committed:
Many minor typo and style fixes in EELSModel docstrings and comments

magnunor committed:
Fix bug EELSModel: enable/disable_free_onset_energy

The functions were previously using the disused component attribute
free_onset_energy, instead of the correct onset_energy.free.

The commit also adds tests:
- i) for checking onset_energy.free is correctly set
- ii) that this actually affects the fitting itself. Ergo, to see
if the onset_energy values changes if enable_free_onset_energy is
run, and to see that the onset_energy values does NOT change after
using disable_free_onset_energy.

magnunor committed:
Fix bug in EELSModel.free_edges

The function was only setting the parameter intensity to free,
not onset_energy and fine_structure_coeff.

This commit fixes this, and adds tests for:

i) checking the free attribute in these parameters are set to the
correct value
ii) Check that this affects the fitting results, by running free_edges
and seeing if the values change after fitting

magnunor committed:
Add testing of several functions in EELSModel

These functions were either not tested, or only partially tested.

magnunor committed:
Add upcoming changes for PR 2865: EELSModel fixes

magnunor commented:
@ericpre, the commit messages should be a bit more descriptive now, especially for the bug-fixes.

ericpre commented on hyperspy/models/eelsmodel.py:
```diff
@@ -794,7 +794,7 @@ def enable_free_onset_energy(self, edges_list=None):
             edges_list = [self._get_component(x) for x in edges_list]
         for edge in edges_list:
             if edge.isbackground is False:
-                edge.free_onset_energy = True
```
Are you sure that `free_onset_energy` is disused? It is used in [`smart_fit`.](https://github.com/hyperspy/hyperspy/blob/1ffb3fab33e607045a37f30c1463350b72617e10/hyperspy/models/eelsmodel.py#L540-L545)

It should be possible to deprecate it and raise a warning when it is used.

magnunor replied:
Initially it seemed like it was some leftover functionality, but now I'm not sure how best to handle it.

I'm guessing it is a bit confusing having both `free_onset_energy` and `onset_energy.free`. In addition, the `onset_energy.free` follows the "parameter" paradime, which is consistent with how the model fitting is done elsewhere?

I guess the `enable_free_onset_energy` function can set both `edge.free_onset_energy = True` and `edge.onset_energy.free = True`.

Then we can add a deprecation warning for `free_onset_energy`, and change `_fit_edge` from

```python
if edge.free_onset_energy is True:
    edge.onset_energy.free = True
    self.fit(**kwargs)
    edge.onset_energy.free = False
```

to 

```python
if edge.onset_energy.free is True:
    self.fit(**kwargs)
    edge.onset_energy.free = False
```

ericpre commented on hyperspy/models/eelsmodel.py:
```diff
@@ -887,6 +887,8 @@ def free_edges(self, edges_list=None):
         for edge in edges_list:
             if edge.isbackground is False:
                 edge.intensity.free = True
```
Use `set_parameters_free` instead?

magnunor replied:
This would set the `effective_angle` parameter also to free, which I don't think would be the intention.

ericpre replied:
Not if you specify the parameter list: https://hyperspy.readthedocs.io/en/latest/api/hyperspy.component.html#hyperspy.component.Component.set_parameters_free

ericpre reviewed (commented):
Thanks, I have left some questions/suggestions.

magnunor committed:
Change EELSModel.free_edges to use set_parameters_free
